cclient/iota_client_extended_api.c

```diff
@@ -56,6 +56,107 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+  iota_transaction_t tx = {};
+  flex_trit_t* tmp_trytes = NULL;
+  flex_trit_t trunk_hash[FLEX_TRIT_SIZE_243];
+  int64_t current_index = 0, last_index = 0;
+  bool is_tail = true;
+
+  log_info(logger_id, "[%s:%d]\n", __func__, __LINE__);
+  get_trytes_req = get_trytes_req_new();
+  if (!get_trytes_req) {
+    ret_code = RC_CCLIENT_OOM;
+    log_error(logger_id, "%s get_trytes_req_new failed: %s\n", __func__,
+              error_2_string(ret_code));
+    goto cleanup;
+  }
+  get_trytes_res = get_trytes_res_new();
+  if (!get_trytes_res) {
+    ret_code = RC_CCLIENT_OOM;
+    log_error(logger_id, "%s get_trytes_res_new failed: %s\n", __func__,
+              error_2_string(ret_code));
+    goto cleanup;
+  }
+
+  do {
+    memcpy(trunk_hash, is_tail ? tail_hash : transaction_trunk(&tx),
+           FLEX_TRIT_SIZE_243);
+    ret_code = hash243_queue_push(&get_trytes_req->hashes, trunk_hash);
+    if (ret_code != RC_OK) {
+      log_error(logger_id, "%s hash243_queue_push failed: %s\n", __func__,
+                error_2_string(ret_code));
+      goto cleanup;
+    }
+    // get trytes from the given hash
+    ret_code = iota_client_get_trytes(serv, get_trytes_req, get_trytes_res);
+    if (ret_code != RC_OK) {
+      log_error(logger_id, "%s iota_client_get_trytes failed: %s\n", __func__,
+                error_2_string(ret_code));
+      goto cleanup;
+    }
+    hash243_queue_pop(&get_trytes_req->hashes);
+    // Get the transaction trytes
+    tmp_trytes = hash8019_queue_peek(get_trytes_res->trytes);
+    if (!tmp_trytes) {
+      ret_code = RC_CCLIENT_RES_ERROR;
+      log_error(logger_id, "%s read transaction trytes failed: %s\n", __func__,
+                error_2_string(ret_code));
+      goto cleanup;
+    }
+    // Create a transaction with the received trytes
+    transaction_deserialize_from_trits(&tx, tmp_trytes, false);
+    transaction_set_hash(&tx, is_tail ? tail_hash : trunk_hash);
+    if (is_tail) {
+      current_index = transaction_current_index(&tx);
+      // Check that the first transaction we get is really a tail transaction
+      // Its index must be 0 and the list of transactions empty.
+      // It is an error if the first transaction has index > 0
+      if (current_index != 0) {
+        ret_code = RC_CCLIENT_INVALID_TAIL_HASH;
+        log_error(logger_id, "%s current index is not zero: %s\n", __func__,
+                  error_2_string(ret_code));
+        goto cleanup;
+      }
+      last_index = transaction_last_index(&tx);
+      is_tail = false;
+    } else {
+      current_index = transaction_current_index(&tx);
+    }
+
+    if (trytes) {
+      // save raw trytes if we want them.
+      hash_array_push(trytes, tmp_trytes);
+    }
+    if (bundle) {
+      // add this transaction to bundle
+      bundle_transactions_add(bundle, &tx);
+    }
+    // The response is not needed anymore
+    hash8019_queue_free(&get_trytes_res->trytes);
+  } while (current_index != last_index);
```

Comment:
Doesn't the JS implementation do additional checks here?
What if the bundle we're trying to traverse is malformed maliciously? e.g. varying indices, ...
Ideally, we should at least check that the `current_index` of the retrieved transaction matches our expected index (`current_index+1`) and maybe also check that bundle hash is consistent

Reply:
yes, JS doesn't have additional checks.
I expect the current_index is in the right order(0,1,2..) and the bundle hash is consistent since transactions are fetched from the Tangle.

Reply:
We fetch via getTrytes and not getBundle; getTrytes does not do any validation of tx validity.

Reply:
done.